thriftpy/parser/parser.py

```diff
@@ -880,3 +301,224 @@ def _get_ttype(inst, default_ttype=None):
+str_val_b = "'" <(('\\\\' "'") | (~"'" anything))*>:val "'" -> val
+Identifier = not_reserved <(Letter | '_') (Letter | Digit | '.' | '_')*>
+identifier_ref :module = Identifier:val -> IdentifierRef(module, val)  # unresolved reference
+annotations = (brk '(' annotation*:name_vals')' brk -> name_vals)? | !(())  # always optional
+annotation = brk Identifier:name brk ('=' brk Literal)?:val brk ListSeparator? brk -> name, val
+ListSeparator = ',' | ';'
+Letter = letter  # parsley built-in
+Digit = digit  # parsley built-in
+Comment = cpp_comment | c_comment | python_comment
+brk = <(' ' | '\t' | '\n' | '\r' | c_comment | cpp_comment | python_comment)*>
+cpp_comment = '//' rest_of_line
+c_comment = '/*' <(~'*/' anything)*>:body '*/' -> body
+python_comment = '#' rest_of_line
+rest_of_line = <('\\\n' | (~'\n' anything))*>
+immutable = '(' brk 'python.immutable' brk '=' brk '""' brk ')'
+Reserved = ('__CLASS__' | '__DIR__' | '__FILE__' | '__FUNCTION__' | '__LINE__' | '__METHOD__' |
+            '__NAMESPACE__' | 'abstract' | 'alias' | 'and' | 'args' | 'as' | 'assert' | 'BEGIN' |
+            'begin' | 'binary' | 'bool' | 'break' | 'byte' | 'case' | 'catch' | 'class' | 'clone' |
+            'const' | 'continue' | 'declare' | 'def' | 'default' | 'del' | 'delete' | 'do' |
+            'double' | 'dynamic' | 'elif' | 'else' | 'elseif' | 'elsif' | 'END' | 'end' |
+            'enddeclare' | 'endfor' | 'endforeach' | 'endif' | 'endswitch' | 'endwhile' | 'ensure' |
+            'enum' | 'except' | 'exception' | 'exec' | 'extends' | 'finally' | 'float' | 'for' |
+            'foreach' | 'from' | 'function' | 'global' | 'goto' | 'i16' | 'i32' | 'i64' | 'if' |
+            'implements' | 'import' | 'in' | 'include' | 'inline' | 'instanceof' | 'interface' |
+            'is' | 'lambda' | 'list' | 'map' | 'module' | 'namespace' | 'native' | 'new' | 'next' |
+            'nil' | 'not' | 'oneway' | 'optional' | 'or' | 'pass' | 'print' | 'private' |
+            'protected' | 'public' | 'public' | 'raise' | 'redo' | 'register' | 'required' |
+            'rescue' | 'retry' | 'return' | 'self' | 'service' | 'set' | 'sizeof' | 'static' |
+            'string' | 'struct' | 'super' | 'switch' | 'synchronized' | 'then' | 'this' |
+            'throw' | 'throws' | 'transient' | 'try' | 'typedef' | 'undef' | 'union' | 'union' |
+            'unless' | 'unsigned' | 'until' | 'use' | 'var' | 'virtual' | 'void' | 'volatile' |
+            'when' | 'while' | 'with' | 'xor' | 'yield')
+not_reserved = ~(Reserved (' ' | '\t' | '\n'))
+'''
+
+
+BASE_TYPE_MAP = {
+    'bool': TType.BOOL,
+    'byte': TType.BYTE,
+    'i8': TType.BYTE,
+    'i16': TType.I16,
+    'i32': TType.I32,
+    'i64': TType.I64,
+    'double': TType.DOUBLE,
+    'string': TType.STRING,
+    'binary': TType.BINARY
+}
+
+
+PARSER = parsley.makeGrammar(
+    GRAMMAR, 
+    {
+        'Document': collections.namedtuple('Document', 'headers definitions'),
```

Comment:
i like the use of namedtuples here, and that parsley supports that sort of set-once/immutable approach